File: src/main/java/no/nav/sbl/dialogarena/mininnboks/provider/rest/henvendelse/HenvendelseController.java
package no.nav.sbl.dialogarena.mininnboks.provider.rest.henvendelse;

import no.nav.common.auth.SubjectHandler;
import no.nav.metrics.Event;
import no.nav.metrics.MetricsFactory;
import no.nav.sbl.dialogarena.mininnboks.consumer.HenvendelseService;
import no.nav.sbl.dialogarena.mininnboks.consumer.domain.*;
import no.nav.sbl.dialogarena.mininnboks.consumer.tilgang.TilgangDTO;
import no.nav.sbl.dialogarena.mininnboks.consumer.tilgang.TilgangService;
import no.nav.tjeneste.domene.brukerdialog.henvendelse.v1.sendinnhenvendelse.meldinger.WSSendInnHenvendelseResponse;
import org.apache.cxf.binding.soap.SoapFault;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.inject.Inject;
import javax.ws.rs.*;
import javax.ws.rs.core.Response;
import javax.xml.ws.soap.SOAPFaultException;
import java.util.*;
import java.util.stream.Collectors;

import static java.util.Optional.empty;
import static java.util.Optional.of;
import static java.util.stream.Collectors.groupingBy;
import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
import static javax.ws.rs.core.Response.Status.CREATED;
import static javax.ws.rs.core.Response.Status.NOT_FOUND;
import static javax.ws.rs.core.Response.Status.NOT_ACCEPTABLE;
import static javax.ws.rs.core.Response.ok;
import static javax.ws.rs.core.Response.status;
import static no.nav.sbl.dialogarena.mininnboks.consumer.domain.Henvendelsetype.SVAR_SBL_INNGAAENDE;
import static no.nav.sbl.dialogarena.mininnboks.consumer.domain.Traad.NYESTE_FORST;

@Path("/traader")
@Produces(APPLICATION_JSON)
public class HenvendelseController {

    final Logger logger = LoggerFactory.getLogger(HenvendelseController.class);

    @Inject
    private HenvendelseService henvendelseService;

    @Inject
    private TilgangService tilgangService;

    @GET
    public List<Traad> hentTraader() {
        String fnr = SubjectHandler.getIdent().orElseThrow(() -> new NotAuthorizedException("Fant ikke brukers OIDC-token"));
        List<Henvendelse> henvendelser = henvendelseService.hentAlleHenvendelser(fnr);

        final Map<String, List<Henvendelse>> traader = henvendelser.stream()
                .collect(groupingBy(henvendelse -> henvendelse.traadId));

        return traader.values().stream()
                .map(this::filtrerDelsvar)
                .map(Traad::new)
                .sorted(NYESTE_FORST)
                .collect(Collectors.toList());
    }

    @GET
    @Path("/{id}")
    public Response hentEnkeltTraad(@PathParam("id") String id) {
        Optional<Traad> optionalTraad = hentTraad(id);
        if (optionalTraad.isPresent()) {
            Traad traad = optionalTraad.get();
            Traad filtrertTraad = new Traad(filtrerDelsvar(traad.meldinger));
            return ok(filtrertTraad).build();
        } else {
            return status(NOT_FOUND.getStatusCode()).build();
        }
    }

    @POST
    @Path("/lest/{behandlingsId}")
    public Response markerSomLest(@PathParam("behandlingsId") String behandlingsId) {
        henvendelseService.merkSomLest(behandlingsId);
        return ok(TupleResultat.of("traadId", behandlingsId)).build();
    }

    @POST
    @Path("/allelest/{behandlingskjedeId}")
    public Response markerAlleSomLest(@PathParam("behandlingskjedeId") String behandlingskjedeId) {
        henvendelseService.merkAlleSomLest(behandlingskjedeId);
        return ok(TupleResultat.of("traadId", behandlingskjedeId)).build();
    }

    @POST
    @Path("/sporsmal")
    @Consumes(APPLICATION_JSON)
    public Response sendSporsmal(Sporsmal sporsmal) {
        String fnr = SubjectHandler.getIdent().orElseThrow(() -> new NotAuthorizedException("Fant ikke brukers OIDC-token"));
        Henvendelse henvendelse = lagHenvendelse(fnr, sporsmal);

        Event metrikk = MetricsFactory.createEvent("mininnboks.sendsporsmal");
        metrikk.addTagToReport("tema", sporsmal.temagruppe);
        metrikk.report();

        WSSendInnHenvendelseResponse response = henvendelseService.stillSporsmal(henvendelse, fnr);

        return status(CREATED).entity(new NyHenvendelseResultat(response.getBehandlingsId())).build();
    }

    @POST
    @Path("/sporsmaldirekte")
    @Consumes(APPLICATION_JSON)
    public Response sendSporsmalDirekte(Sporsmal sporsmal) {
        String fnr = SubjectHandler.getIdent().orElseThrow(() -> new NotAuthorizedException("Fant ikke brukers OIDC-token"));
        Henvendelse henvendelse = lagHenvendelse(fnr, sporsmal);

        Event metrikk = MetricsFactory.createEvent("mininnboks.sendsporsmaldirekte");
        metrikk.addTagToReport("tema", sporsmal.temagruppe);
        metrikk.report();

        WSSendInnHenvendelseResponse response = henvendelseService.stillSporsmalDirekte(henvendelse, fnr);

        return status(CREATED).entity(new NyHenvendelseResultat(response.getBehandlingsId())).build();
    }

    @POST
    @Path("/svar")
    @Consumes(APPLICATION_JSON)
    public Response sendSvar(Svar svar) {
        assertFritekst(svar.fritekst);
        Optional<Traad> traadOptional = hentTraad(svar.traadId);
        if (!traadOptional.isPresent()) {
            return status(NOT_FOUND.getStatusCode()).build();
        }

        Traad traad = traadOptional.get();

        if (!traad.kanBesvares) {
            return status(NOT_ACCEPTABLE.getStatusCode()).build();
        }

        Henvendelse henvendelse = new Henvendelse(svar.fritekst, traad.nyeste.temagruppe);
        henvendelse.traadId = svar.traadId;
        henvendelse.eksternAktor = traad.nyeste.eksternAktor;
        henvendelse.brukersEnhet = traad.eldste.brukersEnhet;
        henvendelse.tilknyttetEnhet = traad.nyeste.tilknyttetEnhet;
        henvendelse.type = SVAR_SBL_INNGAAENDE;
        henvendelse.opprettet = new Date();
        henvendelse.markerSomLest();
        henvendelse.erTilknyttetAnsatt = traad.nyeste.erTilknyttetAnsatt;
        henvendelse.kontorsperreEnhet = traad.nyeste.kontorsperreEnhet;

        Event metrikk = MetricsFactory.createEvent("mininnboks.sendsvar");
        metrikk.addTagToReport("tema", traad.nyeste.temaKode);
        metrikk.report();

        String fnr = SubjectHandler.getIdent().orElseThrow(() -> new NotAuthorizedException("Fant ikke brukers OIDC-token"));
        WSSendInnHenvendelseResponse response = henvendelseService.sendSvar(henvendelse, fnr);

        return status(CREATED).entity(new NyHenvendelseResultat(response.getBehandlingsId())).build();
    }

    private Henvendelse lagHenvendelse(String fnr, Sporsmal sporsmal) {
        Temagruppe temagruppe = Temagruppe.valueOf(sporsmal.temagruppe);
        assertFritekst(sporsmal.fritekst);
        assertTemagruppeTilgang(fnr, temagruppe);

        return new Henvendelse(sporsmal.fritekst, temagruppe);
    }

    private static void assertFritekst(String fritekst) {
        if (fritekst == null) {
            throw new BadRequestException("Fritekst må være sendt med");
        } else if (fritekst.trim().length() == 0) {
            throw new BadRequestException("Fritekst må inneholde tekst");
        } else if (fritekst.trim().length() > 1000) {
            throw new BadRequestException("Fritekst kan ikke være lengre enn 1000 tegn");
        }
    }

    private void assertTemagruppeTilgang(String fnr, Temagruppe temagruppe) {
        if (!Temagruppe.GODKJENTE_FOR_INNGAAENDE_SPORSMAAL.contains(temagruppe)) {
            throw new BadRequestException("Innsending på temagruppe er ikke godkjent");
        }
        if (temagruppe == Temagruppe.OKSOS && !harTilgangTilKommunalInnsending(fnr)) {
            throw new BadRequestException("Bruker har ikke lov til å sende inn henvendelse på temagruppe OKSOS.");
        }
    }

    private boolean harTilgangTilKommunalInnsending(String fnr) {
        return TilgangDTO.Resultat.OK.equals(tilgangService.harTilgangTilKommunalInnsending(fnr).getResultat());
    }

    private List<Henvendelse> filtrerDelsvar(List<Henvendelse> traad) {
        if (traadHarIkkeSkriftligSvarFraNAV(traad)) {
            return traad.stream()
                    .filter(henvendelse -> henvendelse.type != Henvendelsetype.DELVIS_SVAR_SKRIFTLIG)
                    .collect(Collectors.toList());
        }
        return traad;
    }

    private boolean traadHarIkkeSkriftligSvarFraNAV(List<Henvendelse> traad) {
        Optional<Henvendelse> skriftligSvarFraNAV = traad.stream()
                .filter(henvendelse -> henvendelse.type == Henvendelsetype.SVAR_SKRIFTLIG)
                .findAny();
        return !skriftligSvarFraNAV.isPresent();
    }

    private Optional<Traad> hentTraad(String id) {
        try {
            List<Henvendelse> meldinger = henvendelseService.hentTraad(id);
            if (meldinger == null || meldinger.isEmpty()) {
                return empty();
            } else {
                return of(new Traad(meldinger));
            }
        } catch (SoapFault | SOAPFaultException fault) {
            logger.error("Fant ikke tråd med id: " + id, fault);
            return empty();
        }
    }

    static final class NyHenvendelseResultat {
        public final String behandlingsId;

        private NyHenvendelseResultat(String behandlingsId) {
            this.behandlingsId = behandlingsId;
        }
    }

    static final class TupleResultat extends HashMap<String, String> {
        public static TupleResultat of(String key, String value) {
            TupleResultat tr = new TupleResultat();
            tr.put(key, value);
            return tr;
        }
    }

}

File: src/main/java/no/nav/sbl/dialogarena/mininnboks/provider/rest/henvendelse/HenvendelseController.kt
package no.nav.sbl.dialogarena.mininnboks.provider.rest.henvendelse

import io.ktor.application.call
import io.ktor.http.HttpStatusCode
import io.ktor.request.receive
import io.ktor.response.respond
import io.ktor.routing.Route
import io.ktor.routing.get
import io.ktor.routing.post
import io.ktor.routing.route
import no.nav.sbl.dialogarena.mininnboks.consumer.HenvendelseService
import no.nav.sbl.dialogarena.mininnboks.consumer.domain.*
import no.nav.sbl.dialogarena.mininnboks.consumer.tilgang.TilgangDTO
import no.nav.sbl.dialogarena.mininnboks.consumer.tilgang.TilgangService
import no.nav.sbl.dialogarena.mininnboks.provider.rest.ubehandletmelding.conditionalAuthenticate
import no.nav.sbl.dialogarena.mininnboks.provider.rest.ubehandletmelding.getIdentifikator
import org.apache.cxf.binding.soap.SoapFault
import java.util.*
import javax.ws.rs.BadRequestException
import javax.ws.rs.NotAuthorizedException


fun Route.HenvendelseController(henvendelseService: HenvendelseService, tilgangService: TilgangService, useAuthentication: Boolean) {

    conditionalAuthenticate(useAuthentication) {
        route("/traader") {

            get("/") {
                val fnr = call.getIdentifikator() ?: throw  NotAuthorizedException("Fant ikke brukers OIDC-token")
                val henvendelser = henvendelseService.hentAlleHenvendelser(fnr)
                val traader = henvendelser!!.groupBy { it?.traadId }
                call.respond(traader.values
                        .map { t: List<Henvendelse?>? -> filtrerDelsvar(t) }
                        .map { meldinger: List<Henvendelse?>? -> Traad(meldinger) }
                        .sortedWith(Traad.NYESTE_FORST))
            }

            get("/{id}") {
                val id = call.parameters.get("id")
                val optionalTraad = hentTraad(henvendelseService, id)
                if (optionalTraad.isPresent) {
                    val traad = optionalTraad.get()
                    val filtrertTraad = Traad(filtrerDelsvar(traad.meldinger))
                    call.respond(filtrertTraad)
                } else {
                    call.respond(HttpStatusCode.Companion.NotFound)
                }
            }

            post("/lest/{behandlingsId}") {
                val behandlingsId = call.parameters.get("behandlingsId")

                if (behandlingsId != null) {
                    henvendelseService.merkSomLest(behandlingsId)
                    call.respond(TupleResultat.of("traadId", behandlingsId))
                }
            }

            post("/allelest/{behandlingskjedeId}") {
                var behandlingskjedeId = call.parameters.get("behandlingskjedeId")
                henvendelseService.merkAlleSomLest(behandlingskjedeId)
                call.respond(TupleResultat.of("traadId", behandlingskjedeId))
            }

            post("/sporsmal") {
                val fnr = call.getIdentifikator() ?: throw  NotAuthorizedException("Fant ikke brukers OIDC-token")
                val sporsmal = call.receive(Sporsmal::class)
                val henvendelse = lagHenvendelse(tilgangService, fnr, sporsmal)
                //  val metrikk = MetricsClient.createEvent("mininnboks.sendsporsmal")
                //  metrikk.addTagToReport("tema", sporsmal.temagruppe)
                // metrikk.report()
                val response = henvendelseService!!.stillSporsmal(henvendelse, fnr)
                call.respond(HttpStatusCode.Created, NyHenvendelseResultat(response.behandlingsId))
            }

            post("/sporsmaldirekte") {
                val fnr = call.getIdentifikator() ?: throw  NotAuthorizedException("Fant ikke brukers OIDC-token")
                val sporsmal = call.receive(Sporsmal::class)
                val henvendelse = lagHenvendelse(tilgangService, fnr, sporsmal)
                //val metrikk = MetricsFactory.createEvent("mininnboks.sendsporsmaldirekte")
                //metrikk.addTagToReport("tema", sporsmal.temagruppe)
                //metrikk.report()
                val response = henvendelseService!!.stillSporsmalDirekte(henvendelse, fnr)
                call.respond(HttpStatusCode.Created, NyHenvendelseResultat(response.behandlingsId))
            }

            post("/svar") {
                val svar = call.receive(Svar::class)
                assertFritekst(svar.fritekst)
                val traadOptional = hentTraad(henvendelseService, svar.traadId)
                if (!traadOptional.isPresent) {
                    call.respond(HttpStatusCode.NotFound)
                }
                val traad = traadOptional.get()
                if (!traad.kanBesvares) {
                    call.respond(HttpStatusCode.NotAcceptable)
                }
                val henvendelse = Henvendelse(svar.fritekst, traad.nyeste?.temagruppe)
                henvendelse.traadId = svar.traadId
                henvendelse.eksternAktor = traad.nyeste?.eksternAktor
                henvendelse.brukersEnhet = traad.eldste?.brukersEnhet
                henvendelse.tilknyttetEnhet = traad.nyeste?.tilknyttetEnhet
                henvendelse.type = Henvendelsetype.SVAR_SBL_INNGAAENDE
                henvendelse.opprettet = Date()
                henvendelse.markerSomLest()
                henvendelse.erTilknyttetAnsatt = traad.nyeste?.erTilknyttetAnsatt
                henvendelse.kontorsperreEnhet = traad.nyeste?.kontorsperreEnhet
                //val metrikk = MetricsFactory.createEvent("mininnboks.sendsvar")
                //metrikk.addTagToReport("tema", traad.nyeste?.temaKode)
                //metrikk.report()
                val fnr = call.getIdentifikator() ?: throw  NotAuthorizedException("Fant ikke brukers OIDC-token")
                val response = henvendelseService!!.sendSvar(henvendelse, fnr)
                call.respond(HttpStatusCode.Created, NyHenvendelseResultat(response!!.behandlingsId))
            }


        }

    }
}

fun lagHenvendelse(tilgangService: TilgangService, fnr: String, sporsmal: Sporsmal): Henvendelse {
    val temagruppe = sporsmal.temagruppe?.let { Temagruppe.valueOf(it) }
    assertFritekst(sporsmal.fritekst)
    if (temagruppe != null) {
        assertTemagruppeTilgang(tilgangService, fnr, temagruppe)
    }
    return Henvendelse(sporsmal.fritekst, temagruppe)
}

fun assertTemagruppeTilgang(tilgangService: TilgangService, fnr: String, temagruppe: Temagruppe) {
    if (!Temagruppe.GODKJENTE_FOR_INNGAAENDE_SPORSMAAL.contains(temagruppe)) {
        throw BadRequestException("Innsending på temagruppe er ikke godkjent")
    }
    if (temagruppe == Temagruppe.OKSOS && !harTilgangTilKommunalInnsending(tilgangService, fnr)) {
        throw BadRequestException("Bruker har ikke lov til å sende inn henvendelse på temagruppe OKSOS.")
    }
}

fun harTilgangTilKommunalInnsending(tilgangService: TilgangService, fnr: String): Boolean {
    return TilgangDTO.Resultat.OK == tilgangService!!.harTilgangTilKommunalInnsending(fnr).resultat
}

fun filtrerDelsvar(traad: List<Henvendelse?>?): List<Henvendelse?>? {
    if (traadHarIkkeSkriftligSvarFraNAV(traad)) {
        return traad?.sortedBy { henvendelse -> henvendelse?.type != Henvendelsetype.DELVIS_SVAR_SKRIFTLIG }
    } else
        return traad
}


fun traadHarIkkeSkriftligSvarFraNAV(traad: List<Henvendelse?>?): Boolean {
    val skriftligSvarFraNAV = traad?.stream()
            ?.filter { henvendelse -> henvendelse?.type == Henvendelsetype.SVAR_SKRIFTLIG }
            ?.findAny()
    return !skriftligSvarFraNAV?.isPresent!!
}

fun hentTraad(henvendelseService: HenvendelseService, id: String?): Optional<Traad> {
    return try {
        val meldinger = henvendelseService.hentTraad(id)
        if (meldinger == null || meldinger.isEmpty()) {
            Optional.empty()
        } else {
            Optional.of(Traad(meldinger))
        }
    } catch (fault: SoapFault) {
       //TODO( logger.error("Fant ikke tråd med id: $id", fault)
        Optional.empty()
    }
}

class NyHenvendelseResultat(val behandlingsId: String?)

class TupleResultat : HashMap<String?, String?>() {
    companion object {
        fun of(key: String, value: String?): TupleResultat {
            val tr = TupleResultat()
            tr[key] = value
            return tr
        }
    }
}


fun assertFritekst(fritekst: String?) {
    if (fritekst == null) {
        throw BadRequestException("Fritekst må være sendt med")
    } else if (fritekst.trim { it <= ' ' }.length == 0) {
        throw BadRequestException("Fritekst må inneholde tekst")
    } else if (fritekst.trim { it <= ' ' }.length > 1000) {
        throw BadRequestException("Fritekst kan ikke være lengre enn 1000 tegn")
    }
}




